_posts/2019-12-02-Nx.md

```diff
@@ -0,0 +1,177 @@
+---
+layout: post
+authors: [dimitri_de_kerf]
+title: 'One repo to rule them all'
+image: /img/2019-12-02-Nx/nx-logo.png
+tags: [Monorepo, Nx, Angular, NestJS]
+category: Architecture
+comments: true
+---
+
+1. [Intro](#intro)
+2. [Monorepo](#monorepo)
+3. [Nx](#nx)
+4. [Full stack applications](#full-stack-applications)
+6. [Conclusion](#conclusion)
+
+# Intro
+
+Imagine yourself working in a large organization with multiple teams, all working on an application that is part of the platform that the organization offers to its clients.
+You get a set of requirements and start implementing your features.
+After setting up your repository and all the tooling, you’re ready to go.
+All goes well during the development process and you’ve come to the end to deliver your hard work.
+After you’ve deployed your application on the platform, you’re introduced with a blank screen and a console log full of errors.
+
+How can this happen?
+Your team followed the requirements and all the contracts that were defined to communicate with several services.
+Maybe these contracts got outdated in the meantime without you knowing it?
+Were there breaking changes introduced?
+Or does the libraries you are using locally differs from the ones used in production and isn’t compatible anymore?
+It’s really difficult for a large organization to manage all the separate projects and to maintain the overview of all of them.
+The complexity will increase and after a while the whole platform becomes unmanageable.
+
+# Monorepo
+
+The problems mentioned above led large software companies to transition the management of their separate projects by bringing them together under one repository: the monorepo.
+Inside the monorepo, you have all the applications working together while you’re developing your app.
+You’ll get feedback immediately when the service you’re calling returns a different answer than expected or when the page you are navigating to is located on another path.
+When you're not sure how a service of the platform is working, you can just hop in the source code inside the repo to find out more about it.
+Code sharing becomes a breeze, because all the code is available in one repository.
+You don’t have to deal anymore with npm packages for shared code in your organization or with the separate configurations, pipelines and versions of the shared libs.
+
+But how can you start setting up a monorepo and what tools are required to do so?
+Having to manage all this code in one repo can be really tedious to set up.
+This is where Nx comes into play.
+
+# Nx
+
+[Nx](https://nx.dev/) provides you with the tools to help you develop into a monorepo.
+Taking their experience, the creators of Nx set up the best practices to structure your repository in a maintainable way.
+Nx is built on top of the Angular CLI, offering you a way to create workspaces and scaffolding out of your apps.
```

Comment:
I would mention that they also support React apps in their repository and that you can even configure your own builders to manage apps from other frameworks